File: src/engine/trigger.rs
use serde::{Deserialize, Serialize};

use crate::engine::state::State;

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Trigger {
    pub(crate) comparator: TriggerComparator,
    pub(crate) target: TriggerTarget,
    pub(crate) value: u32,
}

impl Trigger {
    pub fn is_satisfied(&self, state: &State) -> bool {
        let actual_value = match self.target {
            TriggerTarget::Population => state.population,
            TriggerTarget::Ecology => state.ecology,
            TriggerTarget::Money => state.money,
        };
        return match self.comparator {
            TriggerComparator::Lt => actual_value < self.value,
            TriggerComparator::Lte => actual_value <= self.value,
            TriggerComparator::Eq => actual_value == self.value,
            TriggerComparator::Gte => actual_value >= self.value,
            TriggerComparator::Gt => actual_value > self.value,
        };
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) enum TriggerComparator {
    Lt,
    Lte,
    Eq,
    Gte,
    Gt,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) enum TriggerTarget {
    Population,
    Ecology,
    Money,
}

File: src/lib/engine/trigger.ts
import type { StateVariable } from "./model";
import type { GameState } from "./state";

export type Comparator = "lt" | "lte" | "eq" | "gte" | "gt";

export class Trigger {
    readonly comparator: Comparator;
    readonly target: StateVariable;
    readonly value: number;

    constructor(comparator: Comparator, target: StateVariable, value: number) {
        this.comparator = comparator;
        this.target = target;
        this.value = value;
    }

    isSatisfiedBy(state: GameState): boolean {
        const actualValue = {
            'population': state.population,
            'ecology': state.ecology,
            'money': state.money,
        }[this.target];

        switch (this.comparator) {
            case 'lt':
                return actualValue < this.value;
            case 'lte':
                return actualValue <= this.value;                
            case 'eq':
                return actualValue === this.value;
            case 'gte':
                return actualValue >= this.value;
            case "gt":
                return actualValue > this.value;
        }
    }
};
